pkg/controller/replicaset/replica_set_test.go

```diff
@@ -1982,6 +2005,97 @@ func TestGetPodsToDelete(t *testing.T) {
 	}
 }
 
+func withPodNameSuffix(pod *v1.Pod, suffix string) *v1.Pod {
+	newPod := pod.DeepCopy()
+	newPod.Name = fmt.Sprintf("%s-%s", pod.Name, suffix)
+	return newPod
+}
+
+func TestComparingSameNodeSpreadOnDeletion(t *testing.T) {
+	labelMap := map[string]string{"name": "foo"}
+	rs := newReplicaSet(6, labelMap)
+
+	tests := []struct {
+		name           string
+		rankFunc       func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks
+		expectedSpread bool
+	}{
+		{
+			name: "sort with the old getPodsRankedByRelatedPodsOnSameNode",
+			rankFunc: func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks {
+				podsOnNode := make(map[string]int)
+				for _, pod := range relatedPods {
+					if controller.IsPodActive(pod) {
+						podsOnNode[pod.Spec.NodeName]++
+					}
+				}
+				ranks := make([]int, len(podsToRank))
+				for i, pod := range podsToRank {
+					ranks[i] = podsOnNode[pod.Spec.NodeName]
+				}
+				return controller.ActivePodsWithRanks{Pods: podsToRank, Rank: ranks, Now: metav1.Now()}
+			},
+			expectedSpread: false,
+		},
+		{
+			name: "sort without extra ranks (just rely on the random timestamp sort)",
+			rankFunc: func(podsToRank, relatedPods []*v1.Pod) controller.ActivePodsWithRanks {
+				return controller.ActivePodsWithRanks{Pods: podsToRank, Rank: make([]int, len(podsToRank)), Now: metav1.Now()}
+			},
+			expectedSpread: false,
+		},
+		{
+			name:           "sort with the new getPodsRankedByRelatedPodsOnSameNode",
+			rankFunc:       getPodsRankedByRelatedPodsOnSameNode,
+			expectedSpread: true,
+		},
+	}
+
+	// base time is a month ago, all pods are mocked to create in last month
+	baseTime := time.Now().Add(-time.Hour * 24 * 7 * 30)
+	for _, test := range tests {
+		// In consequence of random sort, each test will run 1000 times.
+		// If 800 times (80%) succeed, we suppose it matches expectation.
+		var succeedCount int
+		for i := 0; i < 1000; i++ {
+			// scheduled, running, ready pods on fake-node-1
+			podAOnNode1 := newPod("pod-a-on-node-1", rs, v1.PodRunning, &metav1.Time{Time: baseTime.Add(time.Hour * time.Duration(rand.Intn(24*7*30)))}, true)
```

Comment:
I think what you had before (fixed dates, close to each other) made more sense as a fair comparison.

We have two competing heuristics:
1. Removing the youngest pods, as they are likely to have less usage 
2. Keeping the spread

My proposal is giving some priority to 1, but when this is equal, it will try to keep 2.

Your proposal is giving priority to 2, but when this is equal, it will to keep 1.

Since your test is giving random timestamps, my proposal wont behave like yours, because it's optimizing for different things. If you keep the timestamps close to each other, they should behave very similar.

Now comes the question: which heuristic is more important, time or spread? I think my proposal gives a good balance between both. Your proposal favors 2 regardless of timestamps.

@kow3ns @soltysh any thoughts?

Reply:
Again, same question.

Reply:
@alculquicondor  Oh, apologies for the late reply. Actually, I have applied `SameNodeRanker` and `SpreadConstraintsRanker` in [CloneSet of OpenKruise](https://openkruise.io/docs/user-manuals/cloneset), to choose Pods on the same node or according to their TopologySpreadConstraints during scale down.

https://github.com/openkruise/kruise/blob/cf95157d707acbfd0fea4a1d850eda57ea7cfe15/pkg/controller/cloneset/sync/cloneset_scale.go#L371-L394

But for Deployment, maybe you are right, it might not need the strict policy to spread pods over nodes for now. The randomized removal works well. So I'm going to hold the PR temporarily and anyone needs this could comment to let us know. WDYT?

Reply:
Yeah, I used to test it without ranks. Turns out the random algorithm could not choose the expected pods that should have been deleted according to SameNode policy.

However, it was tested in my tiny kind cluster. The random algorithm might work better in production cluster, in which pods should be more randomize.

Reply:
> If you keep the timestamps close to each other, they should behave very similar.

Emm.. I can do this in unit-test. But how can we ensure the timestamps of all pods are close in production? The pods were probably scaled up in different time and by different ppl, e.g., manually, HPA.

Reply:
They might not be close in production. And that's when the age criteria will take precedence.

But when all conditions are equal, it would will be completely random.

Note that the timestamps are compared in a logarithmic fashion. Pod A is only consider older than Pod B if it has been running for at least twice the time. Otherwise they are considered equal.

Reply:
I can take an example that we have generally observed in production.

Once there is an application that has deployed for some days. It has several pods scheduled on the same node because of the limited resource when it scaled up.
Now we add some new nodes into the cluster and we have to migrate the colocated pods to the new nodes. So we are going to increase the `replicas` to create several pods, wait them to be ready, and then decrease the `replicas` and hoping Deployment delete the colocated pods. However, it will probably delete the pods we have just created.

You may ask why not just delete the colocated pods and let controller create new pods and schedule to the new nodes. It is because we have to guarantee the number of ready pods to protect availability for applications.

Reply:
I think a migration use case needs an explicit API, rather than toying with the number of replicas.

In a way, it is similar to a rollout, except that in this case you are upgrading/stopping a node.

The first step is cordoning a node. But then you want the replicaset controller to create a new pod before the old pod is deleted.

Can you join the next SIG meeting to discuss this?

Reply:
> Can you join the next SIG meeting to discuss this?

Sure, I would like to. Besides, this case is just one of the various conditions in production. My point is the timestamps of pods are probably complicated and may be far more than "twice the time". So as you said, what we have to choose is keeping the spread strictly or removing the younger pods which may possibly keep the spread in some cases.